src/storage/txn/commands/check_secondary_locks.rs

```diff
@@ -38,8 +42,74 @@ impl CommandExt for CheckSecondaryLocks {
+        &mut self,
+        txn: &mut MvccTxn<S>,
+        key: &Key,
+    ) -> MvccResult<(SecondaryLockStatus, Option<ReleasedLock>)> {
+        let mut released_lock = None;
+        let (status, need_rollback, rollback_overlapped_write) = match txn.reader.load_lock(&key)? {
+            Some(lock) if lock.ts == self.start_ts => {
+                if lock.lock_type == LockType::Pessimistic {
+                    released_lock = txn.unlock_key(key.clone(), true);
+                    let overlapped_write = txn
+                        .reader
+                        .get_txn_commit_record(&key, self.start_ts)?
+                        .unwrap_none();
+                    (SecondaryLockStatus::RolledBack, true, overlapped_write)
+                } else {
+                    (SecondaryLockStatus::Locked(lock), false, None)
+                }
+            }
+            _ => match txn.reader.get_txn_commit_record(&key, self.start_ts)? {
+                TxnCommitRecord::SingleRecord { commit_ts, write } => {
+                    let status = if write.write_type != WriteType::Rollback {
+                        SecondaryLockStatus::Committed(commit_ts)
+                    } else {
+                        SecondaryLockStatus::RolledBack
+                    };
+                    // We needn't write a rollback once there is a write record for it:
+                    // If it's a committed record, it cannot be changed.
+                    // If it's a rollback record, it either comes from another check_secondary_lock
+                    // (thus protected) or the client stops commit actively. So we don't need
+                    // to make it protected again.
+                    (status, false, None)
+                }
+                TxnCommitRecord::OverlappedRollback { .. } => {
+                    (SecondaryLockStatus::RolledBack, false, None)
+                }
+                TxnCommitRecord::None { overlapped_write } => {
+                    (SecondaryLockStatus::RolledBack, true, overlapped_write)
+                }
+            },
+        };
+        if need_rollback {
+            // We must protect this rollback in case this rollback is collapsed and a stale
+            // acquire_pessimistic_lock and prewrite succeed again.
+            if let Some(write) = make_rollback(self.start_ts, true, rollback_overlapped_write) {
+                txn.put_write(key.clone(), self.start_ts, write.as_ref().to_bytes());
+                if txn.collapse_rollback {
+                    txn.collapse_prev_rollback(key.clone())?;
+                }
+            }
+        }
+        Ok((status, released_lock))
+    }
+}
+
 impl<S: Snapshot, L: LockManager> WriteCommand<S, L> for CheckSecondaryLocks {
-    fn process_write(self, snapshot: S, context: WriteContext<'_, L>) -> Result<WriteResult> {
+    fn process_write(mut self, snapshot: S, context: WriteContext<'_, L>) -> Result<WriteResult> {
         let mut txn = MvccTxn::new(
```

Comment:
It seems for every `WriteCommand`, we need to construct a `MvccTxn` for it, maybe we can do this in the future:
```rust
pub(super) trait WriteCommand<S: Snapshot, L: LockManager>: CommandExt {
    fn process_write(self, snapshot: S, context: WriteContext<'_, L>) -> Result<WriteResult> {
        let mut txn = xxx;
        self.process_write_impl(&mut txn, snapshot, context)
    }
    fn process_write_impl(self, txn: &mut MvccTxn, snapshot: S, context: WriteContext<'_, L>) -> Result<WriteResult>;
}
```

Reply:
Maybe the caller can create the transaction and pass it in?

Reply:
> Maybe the caller can create the transaction and pass it in?

I'm afraid that is not possible for some commands, the caller may not have enough information(eg. `Commit::lock_ts`) to construct the `txn`.

Reply:
Yeah, I was thinking we'd need to expose something like that. There is already a `timestamp` method or something, maybe it is good enough?

Reply:
> Yeah, I was thinking we'd need to expose something like that. There is already a `timestamp` method or something, maybe it is good enough?

`Command::ts` use different timestamps for some of the commands, so we need a separate method.